cmd/boulder-migrator/main.go

```diff
@@ -0,0 +1,329 @@
+
+package main
+
+import (
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/codegangsta/cli"
+	_ "github.com/letsencrypt/boulder/Godeps/_workspace/src/github.com/go-sql-driver/mysql"
+
+	"github.com/letsencrypt/boulder/cmd"
+)
+
+type Migration struct {
+	Id      int
+	Desc    string   // Basic explanation of what the migration does
+	UpSQL   []string
+	DownSQL []string
+}
+
+func getYN(question string) (answer bool) {
+	fmt.Printf("\n%s? [y/n]", question)
+	var free bool
+	for {
+		var resp string
+		fmt.Printf("\n# ")
+		_, err := fmt.Scanf("%s", &resp)
+		if err != nil {
+			fmt.Printf("%s: Please enter y or n\n", err)
+			continue
+		}
+		switch resp {
+		case "yes":
+			fallthrough
+		case "YES":
+			fallthrough
+		case "Yes":
+			fallthrough
+		case "Y":
+			fallthrough
+		case "y":
+			answer = true
+			free = true
+		case "no":
+			fallthrough
+		case "NO":
+			fallthrough
+		case "No":
+			fallthrough
+		case "N":
+			fallthrough
+		case "n":
+			answer = false
+			free = true
+		default:
+			fmt.Println("Please enter y or n")
+		}
+		if free {
+			break
+		}
+	}
+	return
+}
+
+func createMigrationTable(Sql *sql.DB) (err error) {
+	tx, err := Sql.Begin()
+	if err != nil {
+		return
+	}
+	_, err = tx.Exec("CREATE TABLE IF NOT EXISTS migrations (id INTEGER, description TEXT, applied DATETIME, down TEXT)")
+	if err != nil {
+		tx.Rollback()
+		return
+	}
+	err = tx.Commit()
+	return
+}
+
+// list applied migrations
+func listMigrations(Sql *sql.DB) (err error) {
+	rows, err := Sql.Query("SELECT id, description, applied FROM migrations")
+	if err != nil {
+		return
+	}
+	defer rows.Close()
+	fmt.Println("# Currently applied migrations\n")
+	for rows.Next() {
+		var id int
+		var desc string
+		var appliedInt interface{}
+		var applied time.Time
+		if err = rows.Scan(&id, &desc, &appliedInt); err != nil {
+			return
+		}
+		applied, ok := appliedInt.(time.Time)
+		if !ok {
+			fmt.Println(ok, applied)
+			return
+		}
+		fmt.Printf("%d: [%s]\n\t%s\n", id, applied, desc)
+	}
+	return
+}
+
+// apply new migration
+func applyMigration(Sql *sql.DB, migration Migration, yes bool) (err error) {
+	fmt.Printf("[migration %d]\n", migration.Id)
+	fmt.Printf("\tDescription: %s\n", migration.Desc)
+	fmt.Printf("\n\tMigrate SQL:\n")
+	for _, sqlLine := range migration.UpSQL {
+		fmt.Printf("\t%s\n", sqlLine)
+	}
+	fmt.Printf("\n\tRevert SQL:\n")
+	for _, sqlLine := range migration.DownSQL {
+		fmt.Printf("\t%s\n", sqlLine)
+	}
+	if !yes {
+		if answer := getYN("\nWould you like to apply this migration"); !answer {
+			fmt.Println("Ok, bye!")
+			return
+		}
+	}
+	tx, err := Sql.Begin()
+	if err != nil {
+		return
+	}
+	for _, sqlStmt := range migration.UpSQL {
+		_, err = tx.Exec(sqlStmt)
+		if err != nil {
+			tx.Rollback()
+			return
+		}
+	}
+	for _, sqlStmt := range migration.DownSQL {
```

Comment:
Oops, didn't see that the output of this loop is used in the next statement. But see comment below.